tamr_unify_client/models/dataset/resource.py

```diff
@@ -94,6 +94,188 @@ def status(self) -> DatasetStatus:
             self.client, status_json, api_path=self.api_path + "/status"
         )
 
+    def from_geo_features(self, features):
+        """Upsert this dataset from a geospatial FeatureCollection or iterable of Features.
+
+        `features` can be:
+
+        - An object that implements ``__geo_interface__`` as a FeatureCollection
+          (see https://gist.github.com/sgillies/2217756)
+        - An iterable of features, where each element is a feature dictionary or an object
+          that implements the ``__geo_interface__`` as a Feature
+        - A map where the "features" key contains an iterable of features
+
+        See: geopandas.GeoDataFrame.from_features()
+
+        :param features: geospatial features
+        """
+        if hasattr(features, "__geo_interface__"):
+            features = features.__geo_interface__
+        if hasattr(features, "get") and features.get("type") == "FeatureCollection":
+            features = features["features"]
+
+        key_attrs = self.key_attribute_names
+        if len(key_attrs) == 1:
+            record_id = "recordId"
+        else:
+            record_id = "compositeRecordId"
+
+        def upsert_stream(features):
+            for feature in features:
+                yield {
+                    "action": "CREATE",
+                    record_id: feature["id"],
+                    "record": self._feature_to_record(
+                        feature, key_attrs, self._geo_attr
+                    ),
+                }
+
+        self.update_records(upsert_stream(features))
+
+    @property
+    def __geo_interface__(self):
+        """Retrieve a representation of this dataset that conforms to the Python Geo Interface.
+
+        Note that this materializes all features; for a streaming interface to features,
+        see :method:`~tamr_unify_client.models.dataset.Dataset.__geo_features__()`
+
+        See https://gist.github.com/sgillies/2217756
+
+        :return: dict[str, object]
+        """
+        return {
+            "type": "FeatureCollection",
+            "features": [feature for feature in self.__geo_features__],
+        }
+
+    @property
+    def __geo_features__(self):
+        """A generator of the records in this dataset represented as geospatial features
+
+        See https://gist.github.com/sgillies/2217756
+
+        :return: stream of features
+        :rtype: Python generator yielding :py:class:`dict[str, object]`
+        """
+        key_attrs = self.key_attribute_names
+        if len(key_attrs) == 1:
+
```

Comment:
style nit: so much whitespace!!!

Reply:
Black style requires all this whitespace.

Reply:
I tried to do it as a lambda, but it was enough harder to read that I went with in-line function definition instead.